### PR TITLE
Fixes #945 and #2420

### LINKED ISSUE #945: MCS returning partial rings with completeRingsOnly=True
_Reported by Luca Fenu_

Here's the thread: http://www.mail-archive.com/rdkit-discuss@lists.sourceforge.net/msg05651.html
Here's the simple example:

```
In [5]: mol3RA = Chem.MolFromSmiles('c1ccc2cc3ccccc3cc2c1')

In [6]: mol3RB = Chem.MolFromSmiles('c1ccc2c(c1)ccc1ccccc21')

In [7]: rdFMCS.FindMCS((mol3RA,mol3RB),completeRingsOnly=True).smartsString
Out[7]: '[#6]1:[#6]:[#6]:[#6]2:[#6](:[#6]:1):[#6]:[#6]:[#6](:[#6]:2):[#6]:[#6]:[#6]:[#6]'
```

### PR DESCRIPTION
**Pull request by greglandrum**
Touching the MCS code is never particularly enjoyable, but it's good to get this one fixed.

### CODE REVIEW

File: Code/GraphMol/FMCS/FMCS.h [hunk 1]
```diff
@@ -19,29 +19,16 @@ namespace RDKit {
 struct MCSParameters;
 
 struct RDKIT_FMCS_EXPORT MCSAtomCompareParameters {
-  bool MatchValences;
-  bool MatchChiralTag;
-  bool MatchFormalCharge;
-  bool RingMatchesRingOnly;
-
- public:
-  MCSAtomCompareParameters()
```
Comment (bp-kelley) [inline Code/GraphMol/FMCS/FMCS.h]: btw I LOVE this feature of c++11, something good they stole from c#

File: Code/GraphMol/FMCS/FMCS.cpp [hunk 2]
```diff
@@ -285,31 +285,9 @@ static bool checkRingMatch(const MCSBondCompareParameters& p, const ROMol& mol1,
...
-        if (ringMatchMatrixSet->isEqual(&br1, &br2,
-                                        &mol2))  // EQUAL Rings found
-          return true;
-      }
-    }
-    // all rings are different
-    return false;
-  } else
-    return true;  // bond1inRing && bond2inRing;   // both bonds are in rings
+
```
Comment (bp-kelley) [inline Code/GraphMol/FMCS/FMCS.cpp]: checkRingMatch ( just checking the logic here )
 true = bonds are the the same (RING or NON RING)
 false = bonds don't match, one RING, one NON RING

File: Code/GraphMol/FMCS/MaximumCommonSubgraph.cpp [hunk 3]
```diff
@@ -456,15 +530,24 @@ bool MaximumCommonSubgraph::growSeeds() {
...
+          possibleMCS = true;
+        } else if (Parameters.MaximizeBonds &&
+                   (fs.getNumBonds() > getMaxNumberBonds() ||
+                    (fs.getNumBonds() == getMaxNumberBonds() &&
+                     fs.getNumAtoms() > getMaxNumberAtoms()))) {
+          possibleMCS = true;
+        }
+        // #945: test here to see if the MCS actually has all rings closed
+        if (possibleMCS && Parameters.BondCompareParameters.CompleteRingsOnly) {
+          possibleMCS = checkIfRingsAreClosed(fs);
```
Comment (bp-kelley) [inline Code/GraphMol/FMCS/MaximumCommonSubgraph.cpp]: We know the MCS is a match between the structures, here just check to see if rings are closed in the match, if not, keep going.

**Comment by greglandrum** [comment]
I tried to figure out how to do this with the BGL code but ended up deciding that it would be quicker to just write the function myself.
Grokking how to use the BGL algorithms for non-trivial stuff is a “bit” of work

**Pull request merged**